feat: support for strict enum mapping

jvzijp opened the pull request:
Closes #325

latonz commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:
```diff
@@ -83,6 +83,18 @@ private static TypeMapping BuildNameMapping(MappingBuilderContext ctx, bool igno
+            .GetMembers()
+            .OfType<IFieldSymbol>();
+
+        if (enumMemberMappings.Count < targetMembers.Count())
```
This does not really detect all unmapped items. Eg. if the source enum contains the values `Value1, Value2, Value3` and the target enum contains the values `Value10`, `Value20`, `Value30` comparing the count does not work. It should work ByValue, ByName and also take the case sensitivity settings into account.

jvzijp replied:
I changed the implementation to look for missing values in source & target enums. All variations "ByValue, ByName and case sensitivity" are taken into account.

latonz commented on src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs:
```diff
@@ -269,4 +269,13 @@ internal static class DiagnosticDescriptors
+    public static readonly DiagnosticDescriptor NotAllEnumValuesMapped = new DiagnosticDescriptor(
+        "RMG034",
+        "Not all enum values in the target enum are mapped",
+        "Not all enum values in the target enum are mapped from {0} to {1}",
```
Instead of one generic diagnostic, I think `Enum member {0} on {1} not found on source enum {2}` and `Enum member {0} on {1} not found on target enum {2}` diagnostics help a lot during troubleshooting. It would also align with how properties diagnostics are reported.

jvzijp replied:
Agreed, fixed now.

latonz reviewed (changes requested):
Thank you for your contribution! I added my feedback 😊

latonz commented on src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs:
```diff
@@ -269,4 +269,13 @@ internal static class DiagnosticDescriptors
+        "Not all enum values in the target enum are mapped",
+        "Not all enum values in the target enum are mapped from {0} to {1}",
+        DiagnosticCategories.Mapper,
+        DiagnosticSeverity.Warning,
```
To not introduce a breaking change, I think the default severity should be Info.

jvzijp replied:
Agreed, fixed now.

jvzijp commented:
> Thank you for your contribution! I added my feedback 😊 Could you also adjust the docs (`01-mapper.md > Strict property mappings` and `04-enum.mdx`)?

I updated `01-mapper.md > Strict property mappings` but I'm not sure what to add to `04-enum.mdx`, what did you have in mind?

latonz reviewed (changes requested):
Thank you for the updates 😊 😊 
I added my feedback and also added a separate comment what I had in mind regarding `04-enum.mdx`.

jvzijp commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:
```diff
@@ -83,6 +107,20 @@ private static TypeMapping BuildNameMapping(MappingBuilderContext ctx, bool igno
             ctx.ReportDiagnostic(DiagnosticDescriptors.EnumNameMappingNoOverlappingValuesFound, ctx.Source, ctx.Target);
         }
 
+        var missingSourceMembers = sourceFieldsByName.Select(item => item.Key).Except(enumMemberMappings.Select(mapping => mapping.Key));
+
+        if (missingSourceMembers.Any())
+        {
+            ctx.ReportDiagnostic(DiagnosticDescriptors.NotAllSourceEnumValuesMapped, missingSourceMembers.First(), ctx.Source, ctx.Target);
+        }
+
+        var missingTargetMembers = targetFieldsByName.Select(item => item.Key).Except(enumMemberMappings.Select(mapping => mapping.Value));
+
+        if (missingTargetMembers.Any())
+        {
+            ctx.ReportDiagnostic(DiagnosticDescriptors.NotAllTargetEnumValuesMapped, missingTargetMembers.First(), ctx.Target, ctx.Source);
+        }
```
A few comments on this:

- Your proposed solution replicates the mapping done earlier with it's own implementation. I think this diagnostic should be based on the actual mapping created. 
- My proposed solution does in fact work for ignoreCase=true since it works with the actual mappings created. 
- I like that your solution reports ALL missing members instead of only the first

latonz replied:
You are totally correct, my bad! I'm also fine with your solution. Probably just change the any to a foreach.

jvzijp commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:
```diff
@@ -51,12 +52,35 @@ private static TypeMapping BuildCastMappingAndDiagnostic(MappingBuilderContext c
+        return BuildEnumToEnumCastMapping(ctx);
+    }
+
+    private static TypeMapping BuildEnumToEnumCastMapping(MappingBuilderContext ctx)
```
The current name implies that it expects both the source and target member are of type Enum.

jvzijp commented on test/Riok.Mapperly.Tests/Mapping/EnumTest.cs:
```diff
@@ -116,7 +122,7 @@ public Task EnumToOtherEnumByNameWithoutOverlap()
         );
 
-        return TestHelper.VerifyGenerator(source);
+        return TestHelper.VerifyGenerator(source, TestHelperOptions.AllowInfoDiagnostics);
```
My bad, removed it.

jvzijp commented on test/Riok.Mapperly.Tests/Mapping/QueryableProjectionEnumTest.cs:
```diff
@@ -48,6 +48,94 @@ TestSourceBuilderOptions.Default with
     }
 
+    [Fact]
+    public void EnumToAnotherEnumByNameMissingTargetValueShouldDiagnostic()
```
All tests are moved to `EnumTest.cs`

jvzijp commented on test/Riok.Mapperly.Tests/Mapping/QueryableProjectionEnumTest.cs:
```diff
@@ -48,6 +48,94 @@ TestSourceBuilderOptions.Default with
+            "B",
+            TestSourceBuilderOptions.Default with
+            {
+                EnumMappingStrategy = EnumMappingStrategy.ByName
```
Added some CaseInsensitive tests.

jvzijp commented on docs/docs/02-configuration/01-mapper.md:
````diff
@@ -91,3 +91,15 @@ set the following two EditorConfig settings (see also [analyzer diagnostics](./1
+[*.cs]
+dotnet_diagnostic.RMG034.severity = error # Unmapped target enum value
+dotnet_diagnostic.RMG035.severity = error # Unmapped source enum value
+```
````
Done.

jvzijp commented on src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs:
```diff
@@ -293,4 +293,22 @@ internal static class DiagnosticDescriptors
+    public static readonly DiagnosticDescriptor NotAllTargetEnumValuesMapped = new DiagnosticDescriptor(
+        "RMG034",
+        "Not all enum values in the target enum are mapped",
+        "Enum member {0} on {1} not found on source enum {2}",
```
Fixed.

jvzijp commented on src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs:
```diff
@@ -293,4 +293,22 @@ internal static class DiagnosticDescriptors
+    public static readonly DiagnosticDescriptor NotAllTargetEnumValuesMapped = new DiagnosticDescriptor(
+        "RMG034",
+        "Not all enum values in the target enum are mapped",
+        "Enum member {0} on {1} not found on source enum {2}",
```
I agree. Fixed.

jvzijp commented on src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs:
```diff
@@ -293,4 +293,22 @@ internal static class DiagnosticDescriptors
         true
     );
+
+    public static readonly DiagnosticDescriptor NotAllTargetEnumValuesMapped = new DiagnosticDescriptor(
```
Done.

jvzijp commented on src/Riok.Mapperly/Diagnostics/DiagnosticDescriptors.cs:
```diff
@@ -293,4 +293,22 @@ internal static class DiagnosticDescriptors
+        true
+    );
+
+    public static readonly DiagnosticDescriptor NotAllSourceEnumValuesMapped = new DiagnosticDescriptor(
```
Done.

latonz commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:
```diff
@@ -78,11 +97,31 @@ private static TypeMapping BuildNameMapping(MappingBuilderContext ctx, bool igno
             .ToDictionary(x => x.Source.Name, x => x.Target!.Name);
 
-        if (enumMemberMappings.Count == 0)
+        if (!enumMemberMappings.Any())
```
Is this change needed? Isn't the count check faster?

jvzijp replied:
Is indeed faster than Any(), it's just my personal preference. I've no issue with changing it back :)

latonz commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumMappingBuilder.cs:
```diff
@@ -78,11 +97,31 @@ private static TypeMapping BuildNameMapping(MappingBuilderContext ctx, bool igno
 
+        var missingSourceMembers = sourceFieldsByName.Where(field => !enumMemberMappings.ContainsKey(field.Key));
+        foreach (var member in missingSourceMembers)
+            ctx.ReportDiagnostic(
```
This code base uses braces for foreach and ifs, except if it is only a return/throw statement.

jvzijp replied:
Fixed.

latonz commented on test/Riok.Mapperly.Tests/Mapping/EnumTest.cs:
```diff
@@ -332,4 +338,61 @@ public void EnumToStringMappingDisabledShouldUseToStringMapping()
+            $"enum D {{ {targetEnumValues} }}"
+        );
+
+        return TestHelper.VerifyGenerator(source, null, enumMappingStrategy, sourceEnumValues, targetEnumValues);
```
Instead of using all values as parameters, I'd use an additional test parameter describing the test case, what it actually tests (eg. `by name with missing source value`).
Be reminded to remove unused snapshot files after updating the snapshot name.

jvzijp replied:
Fixed.

latonz commented on test/Riok.Mapperly.Tests/TestHelper.cs:
```diff
@@ -7,10 +7,15 @@ namespace Riok.Mapperly.Tests;
+        var verify = Verify(driver);
+
+        if (args.Any())
+            verify.UseParameters(args);
```
This code base uses braces for ifs except if it is a simple return/throw statement.

jvzijp replied:
Fixed.

latonz reviewed (changes requested):
Thanks for the updates.
There are still a few linter issues...

jvzijp commented:
How do I fix the conventional commits check?

latonz commented:
@jvzijp no need to make this check green since it is not a required one. It's just a reminder to adjust the squashed commit message when merging the PR as needed.
Thank you for the contribution 😊

jvzijp commented:
> 

Ah I see. Great :) My pleasure, it was a great learning experience.
